### PR TITLE
 [Security] add password rehashing capabilities 

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31139
| License       | MIT
| Doc PR        | -
| Maker PR | symfony/maker-bundle#389

This PR adds a new `PasswordUpgraderInterface` to migrate passwords to better hash algos when users log in.

The interface is implemented on relevant user-encoder in core. When Users are managed via Doctrine, the interface should be implemented on the app's UserRepository. This means the feature is opt-in. For new projects, symfony/maker-bundle#389 generates user repositories that provide the needed implementation, so that this feature works by default. Existing projects should add [the new method](https://github.com/symfony/maker-bundle/pull/389/files#diff-5abcba85d436f4dc2698f51b2ed99b91) to their repo.

Because it was needed, Guard's `AuthenticatorInterface::checkCredentials()` method now takes a `PasswordUpgraderInterface` as 3rd argument.

On the validation side, a new `MigratingPasswordEncoder` allows validating a hash using several algos.

TL;DR, this should provide state of the art password management, with an authentication layer that can validate old password hashes and turn them into new ones.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
@weaverryan this would impact maker bundle, we should figure out how.

**Review by linaori — COMMENTED** [review]
I think this is a good idea! I'm not too sure about the opt-in vs opt-out of #30955. The current implementation in this PR is really nice and I prefer it over an event listener. The issue I see here, is that the functionality is shifted from something you can hook into every authenticator via an event, to you must implement this in your authenticator for it to work. This means people will have to wait for upstream changes for every implementation. In this PR I only see it in the Dao authentication provider, meaning every other built-in is still lacking this feature.

TL;DR
While I really like the idea, I think it's something harder to -implement for/get control over by- vendors and applications

File: src/Symfony/Component/Security/Core/User/UserProviderInterface.php [hunk 1]
```diff
@@ -30,6 +30,8 @@
  * @see UserInterface
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @method bool upgradePassword(UserInterface $user, string $newEncodedPassword)
```
Comment (stof) [inline src/Symfony/Component/Security/Core/User/UserProviderInterface.php]: I don't think we should make it part of the UserProviderInterface. There are lots of valid use cases for *not* dealing with a password in your User (see the discussions from the EU-FOSSA hackathon about that). It should be a separate optional interface implemented by the provider IMO.

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function needsRehash(string $encoded): bool
+    {
+        return $this->bestEncoder->needsRehash($encoded);
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: this implementation is not enough. We also need to rehash if any *other* encoder than the best one was actually validating it.

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        if ($this->bestEncoder->isPasswordValid($encoded, $raw, $salt)) {
+            return true;
+        }
+
+        foreach ($this->extraEncoders as $encoder) {
+            if ($encoder->isPasswordValid($encoded, $raw, $salt)) {
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: wouldn't this cause issues in case the password hash is not sufficient to validate things because it does not embed the info about which algo was used ? Modern APIs (`password_hash` and `sodium`) are embedding that info, but older ones (see the MessageDigestPasswordEncoder for instance) are not.

I think allowing any encoder to try to validate a password hashed by another encoder could create a security risk, depending on what the encoder is doing. Especially if you include `plaintext` in the extra encoders (as that means that leaking a argon2id hash is now as bad as storing password in cleartext as they will be accepted by this validation).
For separate encoders, we should keep the logic relying on EncoderAwareInterface to choose a legacy encoder (but then, we should build a way to upgrade that to the best encoder).

Comment (stof) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: and yes, I know that always embedded the algo in the password hash would be much better, but that was not the common way years ago when we started Symfony 2.0

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/User/UserProviderInterface.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Component/Security/Core/User/UserProviderInterface.php]: This was also my first impression. Next easiest thing to try is to make it an optional interface that a user provider *may* implement - an extra super-power. I find user providers confusing in general, but the "user provider system" indeed seems like the proper place for this at this time.

**Comment by weaverryan** [comment]
Description updated with a link to an issue on MakerBundle. Changes would be minimal there, but there would be some. Thanks for thinking about that.

**Comment by nicolas-grekas** [comment]
Thanks for the comments, continuing my experiment (I still didn't run the code yet :) ):
- maker bundle PR submitted, see https://github.com/symfony/maker-bundle/pull/389
- `upgradePassword()` is not split in a dedicated `PasswordUpgraderInterface`
- applicable user providers now all implement `PasswordUpgraderInterface`
- Guard's `AuthenticatorInterface::checkCredentials()` method now takes a `PasswordUpgraderInterface` as 3rd argument

If anyone is courageous enough to give this a try, please do :)

File: src/Symfony/Component/Security/Core/User/UserProviderInterface.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/UserProviderInterface.php]: fixed by moving the method to a new `PasswordUpgraderInterface`

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: I don't understand why. To me, only the best encoder should be used. Can you elaborate?

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: isPasswordValid do validates the password by using whatever algo each encoder is using.
This means using it is sufficient to me, as we already trust it to do authentication.
I added a note about the plain text encoder.

Comment (stof) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: my issue here is that you try *all* encoders here, while old-gen implementations cannot recognizes where the hash is one using their algo, and so could accept password for a hash that is not theirs if they can generate such hash (which is why the PainTextEncoder is a huge issue, as it can generate anything as hash as it returns the input).

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: All encoder implementations we have are safe in this regard. Past these, it's not our responsibility.
I added a warning about using `PlaintextPasswordEncoder` in the docblock of `ChainPasswordEncoder`.

Comment (stof) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: is MessageDigestPasswordEncoder actually safe in this regard ?

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+4.4.0
+-----
+
+ * Added `ChainPasswordEncoder`
+ * Added method `PasswordEncoderInterface::needsRehash()`
+ * Added and implemented `PasswordUpgraderInterface`
+ * Deprecated `BCryptPasswordEncoder`, use `NativePasswordEncoder` instead
```
Comment (stof) [inline src/Symfony/Component/Security/CHANGELOG.md]: this last line looks wrong. Deprecating it is already done in 4.3 AFAICT

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: yep, because it hashes the password and compares the hashes. Providing a non supported hash cannot collide with the password's hash.

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: would this encoder return `true` in this method if the hash was not generated by this encoder but by one of the extra encoders ? Or would it fail entirely or return `false` ?
There is 2 reasons to rehash a password:

- the best encoder was already used, but it now has a stronger hashing and wants to rehash the password
- a weaker encoder was used. In this case, we want to rehash, but this code expects the best encoder to detect that case when passing it the hash generated by the other encoder (and so potentially in a different format).

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: > would this encoder return true in this method if the hash was not generated by this encoder but by one of the extra encoders

Yes, absolutely! There cannot be any other return value in this situation. The encoder must return `true` only if it supports the hash of course - that's what implementations do also.

File: src/Symfony/Component/Security/CHANGELOG.md (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/CHANGELOG.md]: fixed

**Comment by nicolas-grekas** [comment]
PR ready for deeper review before I invest more time in adding tests and/or split in several PRs: the code together with symfony/maker-bundle#389 works well now.

File: src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php [hunk 5]
```diff
@@ -107,6 +108,22 @@ public function supportsClass($class)
...
+     */
+    public function upgradePassword(UserInterface $user, string $newEncodedPassword): void
+    {
+        $class = $this->getClass();
+        if (!$user instanceof $class) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', \get_class($user)));
+        }
+
+        $repository = $this->getRepository();
+        if ($repository instanceof PasswordUpgraderInterface) {
```
Comment (rpkamp) [inline src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php]: A silent failure seems odd here. Maybe log something in debug that the user will not be saved because the repository doesn't support it? At least that way when people are trying to figure out why the feature doesn't appear to be working there is something in the logs to tell them why.

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php [hunk 6]
```diff
@@ -54,9 +55,15 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
...
 
-            if (!$this->encoderFactory->getEncoder($user)->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
+            $encoder = $this->encoderFactory->getEncoder($user);
+
+            if (!$encoder->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
                 throw new BadCredentialsException('The presented password is invalid.');
             }
+
+            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
+                $this->userProvider->upgradePassword($user, $encoder->encodePassword($presentedPassword, $user->getSalt()));
```
Comment (rpkamp) [inline src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: This gives this method two responsibilies: checking the password and upgrading it if needed. Maybe this should be done at some higher -orchistration- level?

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Hashes passwords using the best available encoder.
+ * Validates them using a chain of encoders.
+ *
+ * /!\ Don't put a PlaintextPasswordEncoder in the list as that'd mean a leaked hash
+ * could be used to authenticate successfully without knowing the cleartext password.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ChainPasswordEncoder extends BasePasswordEncoder implements SelfSaltingEncoderInterface
```
Comment (rpkamp) [inline src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: This name is bit misleading to me. It sounds like multiple encoders are asked to encode the password, but there is actually just one that will encode the password, but multiple (the chain) that will state if the password is valid.

File: src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php [hunk 8]
```diff
@@ -90,6 +90,19 @@ public function supportsClass($class)
...
+     * {@inheritdoc}
+     */
+    public function upgradePassword(UserInterface $user, string $newEncodedPassword): void
+    {
+        if (!$user instanceof User) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', \get_class($user)));
+        }
+
+        $this->getUser($user->getUsername())->setPassword($newEncodedPassword);
+        $user->setPassword($newEncodedPassword);
```
Comment (rpkamp) [inline src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: Shouldn't this just throw an exception? Changing the password of an in-memory doesn't make sense as it be reverted back to the hardcoded value on the next request.

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php [hunk 9]
```diff
@@ -104,4 +104,16 @@ public function supportsClass($class)
...
         return false;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function upgradePassword(UserInterface $user, string $newEncodedPassword): void
+    {
+        foreach ($this->providers as $provider) {
+            if ($provider instanceof PasswordUpgraderInterface) {
```
Comment (Simperfit) [inline src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: The same occurs here, should it really fails silently ?

File: src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php]: A hard failure would be even more odd: you don't want your users not being able to log in when password upgrades cannot happen. They must be opportunistic to me. Logging why not!

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: You mean the checkAuthentication method? I've no better idea, can you elaborate?

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: Chaining logic belongs to each domain. Here, the practice of using several encoders is not documented in the literature: it would give a false sense of security when a single good algo is the correct bullet.
I'm fine with the current name. Do we need another name? If yes which one?

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php (hunk 9 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: yes it should: you don't want your users not being able to log in when upgrades cannot happen.

File: src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php (hunk 8 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: what about tests?

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: Upgrades are opportunistic, so this is fine also for static credentials to me.

File: src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php (hunk 5 above)
Comment (rpkamp) [reply src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php]: I agree a hard failure would be incorrect here.

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 6 above)
Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: Yes, the `checkAuthentication` method.
So instead of doing this inside the `checkAuthentication` do this in the code that calls `checkAuthentication`, so move it one level up to decouple the behaviours:

```
if ($provider->checkAuthentication(...)) {
    $provider->upgradePassword(...);
}
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: I'll have a look but I think what misses in your example is the encoder: the "if" must involve it before calling upgrade. Yet the encoder is internal detail of the provider.

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 7 above)
Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: I think we're running into this problem because ideally the encoder and the validator would be separate interfaces (Interface Seggregation principle, one for encoding, another for validation), but pragmatically it's nicer to couple them into one for symmetry.

Anyway, maybe something like `BackwardCompatiblePasswordEncoder`? Or `OldPasswordSchemeAwarePasswordEncoder`?

File: src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php (hunk 8 above)
Comment (rpkamp) [reply src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: > what about tests?

What about them? Having something in the production code just because you want to test it is putting the cart in front of the horse if you ask me.

My main gripe with this is that the API suggests now that the password of in-memory users can be changed, and junior developers with minimal experience might think this actually works. I don't want to imagine the rabbit hole they fall into when they find it isn't working as they expected ...

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php [hunk 10]
```diff
@@ -104,4 +104,16 @@ public function supportsClass($class)
...
+                $provider->upgradePassword($user, $newEncodedPassword);
```
Comment (rpkamp) [inline src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: This should catch `UnsupportedUserException` and try the next one when this happens. Otherwise if the user is coming from the second provider in the chain, but we ask the first one to upgrade the password first it might throw that exception and the whole thing would fail.

Ideally the interface would be extended with `canUpgradePasswordOf(UserInterface $u): bool` so you check upfront which provider is able to this for you instead of trying fist-first.

```
try {
   $provider->upgradePassword($user, $newEncodedPassword);
} catch (UnsupportedUserException) {
   // try the next one
}
```

File: src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php (hunk 8 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: Agreed, interface removed from the class.

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php (hunk 10 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: done thanks

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: > encoder is internal detail of the provider

I confirm - and so is the cleartext password btw. Same for guard btw.
Dead end to me.

File: src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php]: I'm not sure anymore about the logger: it would be noisy for ppl that just don't want to implement password upgrades for some reason.
Better document this: if you need pwd upgrade, you must implement the interface.

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 7 above)
Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: That would suggest that out of all the `PasswordEncoder`s, this one is the best, which is not what we're trying to say :wink:

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: Yes it is :)
It's the best because it enables migrations, and these are the best :)

Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: But what if we come up with another idea, and that is actually better, what will we call that? Can't call it the best, that's already taken, but what is better than the best? See where I'm going with this? :wink:

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 6 above)
Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: Seems to me it could be moved to `\Symfony\Component\Security\Core\Authentication\Provider\UserAuthenticationProvider`. That way it's part of the template method so all AuthenticationProviders can benefit from it, instead of being part of a particular implementation.

```
public function authenticate(TokenInterface $token)
{
    // ...
    try {
        $this->userChecker->checkPreAuth($user);
        $this->checkAuthentication($user, $token);
        if ($this instanceof PasswordUpgradingAuthenticationProvider) {
            // optional: if ($this->needsPasswordRehash($user, $token)) {
            $this->upgradeUserPassword($user, $token);
            // optional: }
        }
        $this->userChecker->checkPostAuth($user);
    } catch (BadCredentialsException $e) {
        if ($this->hideUserNotFoundExceptions) {
            throw new BadCredentialsException('Bad credentials.', 0, $e);
        }
        throw $e;
    }
```

and then `\Symfony\Component\Security\Core\Authentication\Provider\DaoAuthenticationProvider` would implement `PasswordUpgradingAuthenticationProvider`:

```
class DaoAuthenticationProvider extends UserAuthenticationProvider implements PasswordUpgradingAuthenticationProvider
{
   // ...
   public function upgradeUserPassword(UserInterface $user, UsernamePasswordToken $token): void
   {
      if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
          $this->userProvider->upgradePassword($user, $encoder->encodePassword($token->getCredentials(), $user->getSalt()));
      }
   }
}
```

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encoder;
+
+/**
+ * Hashes passwords using the best available encoder.
+ * Validates them using a chain of encoders.
+ *
+ * /!\ Don't put a PlaintextPasswordEncoder in the list as that'd mean a leaked hash
```
Comment (rpkamp) [inline src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: Instead of (or: in addition to) this warning, why not put a check in the constructor that throws an exception when the `PlaintextPasswordEncoder` is passed?

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: MigratingPasswordEncoder

Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: I like that :+1:

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 11 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: I wouldn't do this: technically one might have a reason to do so, we shouldn't close the door.

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: > optional: if ($this->needsPasswordRehash($user, $token)) {

yes but no :) this is not optional: this is part of the critical lifecycle of password migrations.
Not all auth providers have a concept of encoders. Only those should know and care about migrations. Said another way, your initial issue was SRP for the method, this proposal is moving the issue somewhere else. Let's say there is no SRP issue in the first place and everything is fine: checkAuthentication is the correct place to wire opportunistic password upgrades to me.

Comment (rpkamp) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: My issue was with CQS, not SRP. My issue with SRP was in the naming of the `ChainPasswordEncoder`.

My issue here is still CQS, and I think it would be solved by my proposal.

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: Sorry it wouldn't, see my objections about encoders. See also the guard interface.
Opportunistic password upgrades don't align to CQS here to me.

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: But please submit a PR on my fork if I'm missing the point :)

**Comment by nicolas-grekas** [comment]
Continued in #31843

File: src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php (hunk 11 above)
Comment (elchris) [reply src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php]: fwiw, I use plaintextpasswordencoder in my test harness.
